File: sdk/typescript/src/bridge.ts
/**
 * Bridge client implementation for the Aether TypeScript SDK.
 *
 * This module provides the BridgeClient class for connecting to the Aether
 * gateway as a cross-workspace message bridge. Bridges relay messages across
 * workspace boundaries and can send to any principal type in any workspace.
 *
 * Unlike workspace-scoped principals (agents, tasks, users), bridges have no
 * workspace field in their identity — allowing them to send to targets in any
 * workspace. Each bridge identity can only have one active connection at a time
 * (Connection = Lock paradigm).
 *
 * @module bridge
 */

import { AetherClient } from "./client.js";
import type { AetherClientOptions } from "./client.js";
import { MessageType } from "./types.js";
import { InvalidArgumentError } from "./errors.js";
import {
  agentTopic,
  bridgeTopic,
  uniqueTaskTopic,
  taskBroadcastTopic,
  userTopic,
  userWorkspaceTopic,
  globalAgentsTopic,
  globalUsersTopic,
} from "./topics.js";

// =============================================================================
// Bridge Client Options
// =============================================================================

/**
 * Configuration options for the BridgeClient.
 */
export interface BridgeClientOptions extends AetherClientOptions {
  /** The bridge implementation type (e.g., "aether-msgbridge", "webhook-bridge"). Required. */
  implementation: string;
  /** The unique specifier for this bridge instance (e.g., "default", "discord-1"). Required. */
  specifier: string;
}

// =============================================================================
// BridgeClient
// =============================================================================

/**
 * Client for connecting to the Aether gateway as a cross-workspace message bridge.
 *
 * Bridges are identified by implementation and specifier with no workspace field,
 * allowing them to operate across workspace boundaries. Each bridge identity can
 * only have one active connection at a time (Connection = Lock paradigm).
 *
 * BridgeClient extends AetherClient and adds bridge-specific functionality:
 * - Identity management (implementation, specifier)
 * - Messaging helpers to any workspace (sendToAgent, sendToUser, sendToTask, broadcast)
 *
 * Topic format: br::{implementation}::{specifier}
 *
 * @example
 * ```typescript
 * import { BridgeClient } from "@scitrera/aether-client";
 *
 * const bridge = new BridgeClient({
 *   address: "localhost:50051",
 *   implementation: "aether-msgbridge",
 *   specifier: "discord-1",
 * });
 *
 * bridge.onMessage((msg) => {
 *   console.log(`Received from ${msg.sourceTopic}:`, msg.payload);
 * });
 *
 * await bridge.connect();
 *
 * // Send to any workspace — bridges are cross-workspace
 * const encoder = new TextEncoder();
 * bridge.sendToAgent("prod", "my-agent", "instance-1", encoder.encode("Hello!"));
 * bridge.sendToUser("alice", "tab-1", encoder.encode("Notification from Discord"));
 * ```
 */
export class BridgeClient extends AetherClient {
  private readonly _implementation: string;
  private readonly _specifier: string;

  /**
   * Creates a new BridgeClient.
   *
   * @param options - Bridge client configuration
   * @throws {@link InvalidArgumentError} if required fields are missing
   */
  constructor(options: BridgeClientOptions) {
    super(options);

    if (!options.implementation) {
      throw new InvalidArgumentError("Implementation is required", "implementation");
    }
    if (!options.specifier) {
      throw new InvalidArgumentError("Specifier is required", "specifier");
    }

    this._implementation = options.implementation;
    this._specifier = options.specifier;
  }

  // ===========================================================================
  // Identity Accessors
  // ===========================================================================

  /** Returns the bridge's implementation type. */
  get implementation(): string {
    return this._implementation;
  }

  /** Returns the bridge's specifier (instance identifier). */
  get specifier(): string {
    return this._specifier;
  }

  /**
   * Returns this bridge's topic address.
   *
   * Format: br::{implementation}::{specifier}
   */
  get topic(): string {
    return bridgeTopic(this._implementation, this._specifier);
  }

  // ===========================================================================
  // Init Message
  // ===========================================================================

  /** @internal */
  protected override _buildInitMessage(): Record<string, unknown> {
    return {
      bridge: {
        implementation: this._implementation,
        specifier: this._specifier,
      },
      credentials: this._credentials,
      resumeSessionId: this._resumeSessionId,
    };
  }

  // ===========================================================================
  // Message Sending Helpers
  // ===========================================================================

  /**
   * Sends a message to a specific agent in any workspace.
   *
   * @param workspace - Target agent's workspace
   * @param implementation - Target agent's implementation type
   * @param specifier - Target agent's specifier
   * @param payload - Message payload (bytes)
   * @param messageType - Message type. Default: Chat
   */
  sendToAgent(
    workspace: string,
    implementation: string,
    specifier: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Opaque,
  ): void {
    this._sendMessage(agentTopic(workspace, implementation, specifier), payload, messageType);
  }

  /**
   * Sends a message to a specific task in any workspace.
   *
   * For unique tasks (with specifier), uses tu::{workspace}::{impl}::{spec} topic.
   * For non-unique tasks (empty specifier), uses tb::{workspace}::{impl} broadcast topic.
   *
   * @param workspace - Target task's workspace
   * @param implementation - Target task's implementation type
   * @param specifier - Target task's specifier (empty for broadcast)
   * @param payload - Message payload (bytes)
   * @param messageType - Message type. Default: Chat
   */
  sendToTask(
    workspace: string,
    implementation: string,
    specifier: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Opaque,
  ): void {
    const topic = specifier
      ? uniqueTaskTopic(workspace, implementation, specifier)
      : taskBroadcastTopic(workspace, implementation);
    this._sendMessage(topic, payload, messageType);
  }

  /**
   * Sends a message to a specific user session.
   *
   * @param userId - Target user's ID
   * @param windowId - Target user's window/session ID
   * @param payload - Message payload (bytes)
   * @param messageType - Message type. Default: Chat
   */
  sendToUser(
    userId: string,
    windowId: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Opaque,
  ): void {
    this._sendMessage(userTopic(userId, windowId), payload, messageType);
  }

  /**
   * Sends a message to a user's workspace scope.
   *
   * This reaches the user regardless of which window/tab they are using.
   *
   * @param userId - Target user's ID
   * @param workspace - Target workspace
   * @param payload - Message payload (bytes)
   * @param messageType - Message type. Default: Chat
   */
  sendToUserWorkspace(
    userId: string,
    workspace: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Opaque,
  ): void {
    this._sendMessage(userWorkspaceTopic(userId, workspace), payload, messageType);
  }

  // ===========================================================================
  // Broadcast Helpers
  // ===========================================================================

  /**
   * Sends a message to all agents in a workspace.
   *
   * @param workspace - Target workspace
   * @param payload - Message payload (bytes)
   * @param messageType - Message type. Default: Chat
   */
  broadcastToAgents(
    workspace: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Opaque,
  ): void {
    this._sendMessage(globalAgentsTopic(workspace), payload, messageType);
  }

  /**
   * Sends a message to all users in a workspace.
   *
   * @param workspace - Target workspace
   * @param payload - Message payload (bytes)
   * @param messageType - Message type. Default: Chat
   */
  broadcastToUsers(
    workspace: string,
    payload: Uint8Array,
    messageType: MessageType = MessageType.Opaque,
  ): void {
    this._sendMessage(globalUsersTopic(workspace), payload, messageType);
  }
}

File: sdk/go/aether/bridge.go
// Package aether bridge client implementation.
//
// This file provides the BridgeClient type for connecting to the Aether gateway
// as a cross-workspace message bridge. Bridges relay messages across workspace
// boundaries using the PrincipalBridge principal type.

package aether

import (
	pb "github.com/scitrera/aether/api/proto"
)

// =============================================================================
// Bridge Client
// =============================================================================

// BridgeClient is a client for connecting to the Aether gateway as a cross-workspace
// message bridge.
//
// Bridges are identified by implementation and specifier (no workspace field),
// allowing them to operate across workspace boundaries. Each bridge identity can
// only have one active connection at a time (Connection = Lock paradigm).
//
// BridgeClient embeds BaseClient and adds bridge-specific functionality:
//   - Identity management (implementation, specifier)
//   - Messaging helpers (SendToAgent, SendToUser, SendToTask, Broadcast)
//
// Example usage:
//
//	client, err := aether.NewBridgeClient(aether.BridgeOptions{
//	    ClientOptions: aether.ClientOptions{
//	        ServerAddr: "localhost:50051",
//	    },
//	    Implementation: "aether-msgbridge",
//	    Specifier:      "instance-1",
//	})
//	if err != nil {
//	    log.Fatal(err)
//	}
//
//	client.OnMessage(func(ctx context.Context, msg *aether.Message) error {
//	    fmt.Printf("Received from %s: %s\n", msg.SourceTopic, msg.Payload)
//	    return nil
//	})
//
//	if err := client.Connect(ctx); err != nil {
//	    log.Fatal(err)
//	}
//
//	// Start the message loop (blocks until disconnection)
//	if err := client.Run(ctx); err != nil {
//	    log.Fatal(err)
//	}
type BridgeClient struct {
	*BaseClient

	// Identity fields
	implementation string
	specifier      string
	credentials    map[string]string
}

// NewBridgeClient creates a new BridgeClient with the given options.
//
// The client is created but not connected. Call Connect() to establish
// the connection to the server.
//
// Returns an error if required options are missing (ServerAddr, Implementation, Specifier).
func NewBridgeClient(opts BridgeOptions) (*BridgeClient, error) {
	// Validate options
	if err := opts.Validate(); err != nil {
		return nil, err
	}

	// Create base client configuration
	cfg := BaseClientConfig{
		ServerAddr:  opts.ServerAddr,
		Connection:  opts.Connection,
		TLS:         opts.TLS,
		Credentials: opts.Credentials,
	}

	// Create base client
	base, err := NewBaseClient(cfg)
	if err != nil {
		return nil, err
	}

	// Create bridge client
	bc := &BridgeClient{
		BaseClient:     base,
		implementation: opts.Implementation,
		specifier:      opts.Specifier,
		credentials:    opts.Credentials,
	}

	// Set the init message builder
	base.initMsgBuilder = bc.buildInitMessage

	return bc, nil
}

// buildInitMessage creates the InitConnection message for bridge identity.
func (c *BridgeClient) buildInitMessage() *pb.InitConnection {
	return &pb.InitConnection{
		ClientType: &pb.InitConnection_Bridge{
			Bridge: &pb.BridgeIdentity{
				Implementation: c.implementation,
				Specifier:      c.specifier,
			},
		},
		Credentials: c.credentials,
	}
}

// =============================================================================
// Identity Accessors
// =============================================================================

// Implementation returns the bridge's implementation type.
func (c *BridgeClient) Implementation() string {
	return c.implementation
}

// Specifier returns the bridge's specifier (instance identifier).
func (c *BridgeClient) Specifier() string {
	return c.specifier
}

// Topic returns this bridge's topic address.
//
// Format: br::{implementation}::{specifier}
func (c *BridgeClient) Topic() string {
	return BridgeTopic(c.implementation, c.specifier)
}

// =============================================================================
// Message Sending Helpers
// =============================================================================

// SendToAgent sends a message to a specific agent.
//
// Parameters:
//   - workspace: Target agent's workspace
//   - implementation: Target agent's implementation type
//   - specifier: Target agent's specifier
//   - payload: Message payload (bytes)
//
// Uses CHAT message type by default. Use SendToAgentWithType for other types.
func (c *BridgeClient) SendToAgent(workspace, implementation, specifier string, payload []byte) error {
	return c.SendToAgentWithType(workspace, implementation, specifier, payload, pb.MessageType_OPAQUE)
}

// SendToAgentWithType sends a message to a specific agent with a custom message type.
func (c *BridgeClient) SendToAgentWithType(workspace, implementation, specifier string, payload []byte, msgType pb.MessageType) error {
	topic := AgentTopic(workspace, implementation, specifier)
	return c.sendMessage(topic, payload, msgType)
}

// SendToTask sends a message to a specific task.
//
// For unique tasks (with specifier):
//   - Uses tu::{workspace}::{implementation}::{specifier} topic
//
// For non-unique tasks (empty specifier):
//   - Uses tb::{workspace}::{implementation} broadcast topic for load balancing
//
// Parameters:
//   - workspace: Target task's workspace
//   - implementation: Target task's implementation type
//   - specifier: Target task's specifier (empty for broadcast to non-unique tasks)
//   - payload: Message payload (bytes)
//
// Uses CHAT message type by default. Use SendToTaskWithType for other types.
func (c *BridgeClient) SendToTask(workspace, implementation, specifier string, payload []byte) error {
	return c.SendToTaskWithType(workspace, implementation, specifier, payload, pb.MessageType_OPAQUE)
}

// SendToTaskWithType sends a message to a specific task with a custom message type.
func (c *BridgeClient) SendToTaskWithType(workspace, implementation, specifier string, payload []byte, msgType pb.MessageType) error {
	var topic string
	if specifier != "" {
		// Unique task
		topic = UniqueTaskTopic(workspace, implementation, specifier)
	} else {
		// Non-unique task broadcast for load balancing
		topic = TaskBroadcastTopic(workspace, implementation)
	}
	return c.sendMessage(topic, payload, msgType)
}

// SendToUser sends a message to a specific user session.
//
// Parameters:
//   - userID: Target user's ID
//   - windowID: Target user's window/session ID
//   - payload: Message payload (bytes)
//
// Uses CHAT message type by default. Use SendToUserWithType for other types.
func (c *BridgeClient) SendToUser(userID, windowID string, payload []byte) error {
	return c.SendToUserWithType(userID, windowID, payload, pb.MessageType_OPAQUE)
}

// SendToUserWithType sends a message to a specific user session with a custom message type.
func (c *BridgeClient) SendToUserWithType(userID, windowID string, payload []byte, msgType pb.MessageType) error {
	topic := UserTopic(userID, windowID)
	return c.sendMessage(topic, payload, msgType)
}

// SendToUserWorkspace sends a message to a user's workspace scope.
//
// This reaches the user regardless of which window/tab they're using.
//
// Parameters:
//   - userID: Target user's ID
//   - workspace: Target workspace
//   - payload: Message payload (bytes)
//
// Uses CHAT message type by default. Use SendToUserWorkspaceWithType for other types.
func (c *BridgeClient) SendToUserWorkspace(userID, workspace string, payload []byte) error {
	return c.SendToUserWorkspaceWithType(userID, workspace, payload, pb.MessageType_OPAQUE)
}

// SendToUserWorkspaceWithType sends a message to a user's workspace scope with a custom message type.
func (c *BridgeClient) SendToUserWorkspaceWithType(userID, workspace string, payload []byte, msgType pb.MessageType) error {
	topic := UserWorkspaceTopic(userID, workspace)
	return c.sendMessage(topic, payload, msgType)
}

// =============================================================================
// Broadcast Helpers
// =============================================================================

// BroadcastToAgents sends a message to all agents in a workspace.
//
// Parameters:
//   - workspace: Target workspace (all agents in this workspace receive the message)
//   - payload: Message payload (bytes)
//
// Uses CHAT message type by default. Use BroadcastToAgentsWithType for other types.
func (c *BridgeClient) BroadcastToAgents(workspace string, payload []byte) error {
	return c.BroadcastToAgentsWithType(workspace, payload, pb.MessageType_OPAQUE)
}

// BroadcastToAgentsWithType sends a message to all agents in a workspace with a custom message type.
func (c *BridgeClient) BroadcastToAgentsWithType(workspace string, payload []byte, msgType pb.MessageType) error {
	topic := GlobalAgentsTopic(workspace)
	return c.sendMessage(topic, payload, msgType)
}

// BroadcastToUsers sends a message to all users in a workspace.
//
// Parameters:
//   - workspace: Target workspace (all users in this workspace receive the message)
//   - payload: Message payload (bytes)
//
// Uses CHAT message type by default. Use BroadcastToUsersWithType for other types.
func (c *BridgeClient) BroadcastToUsers(workspace string, payload []byte) error {
	return c.BroadcastToUsersWithType(workspace, payload, pb.MessageType_OPAQUE)
}

// BroadcastToUsersWithType sends a message to all users in a workspace with a custom message type.
func (c *BridgeClient) BroadcastToUsersWithType(workspace string, payload []byte, msgType pb.MessageType) error {
	topic := GlobalUsersTopic(workspace)
	return c.sendMessage(topic, payload, msgType)
}

// =============================================================================
// Generic Message Sending
// =============================================================================

// SendMessage sends a message to a specified topic.
//
// This is the low-level sending method. For most use cases, prefer the
// type-specific helpers (SendToAgent, SendToUser, SendToTask, etc.).
//
// Parameters:
//   - targetTopic: The destination topic string
//   - payload: Message payload (bytes)
//   - msgType: Message type (CHAT, CONTROL, TOOL_CALL, EVENT, METRIC)
func (c *BridgeClient) SendMessage(targetTopic string, payload []byte, msgType pb.MessageType) error {
	return c.sendMessage(targetTopic, payload, msgType)
}

// SendChatMessage sends a CHAT message to a specified topic.
func (c *BridgeClient) SendChatMessage(targetTopic string, payload []byte) error {
	return c.sendMessage(targetTopic, payload, pb.MessageType_OPAQUE)
}

// SendControlMessage sends a CONTROL message to a specified topic.
func (c *BridgeClient) SendControlMessage(targetTopic string, payload []byte) error {
	return c.sendMessage(targetTopic, payload, pb.MessageType_CONTROL)
}

// SendToolCallMessage sends a TOOL_CALL message to a specified topic.
func (c *BridgeClient) SendToolCallMessage(targetTopic string, payload []byte) error {
	return c.sendMessage(targetTopic, payload, pb.MessageType_TOOL_CALL)
}
